DigitalLearningSolutions.Web.Tests/ControllerHelpers/DateValidatorTests.cs

```diff
@@ -1,149 +1,250 @@
+        [TestCase(1, 13, 3000, false, true, false, "month")]
+        [TestCase(1, 1, 0, false, false, true, "year")]
+        [TestCase(1, 1, 10000, false, false, true, "year")]
+        [TestCase(0, 0, 3000, true, true, false, "day and month")]
+        [TestCase(0, 1, 0, true, false, true, "day and year")]
+        [TestCase(1, 0, 0, false, true, true, "month and year")]
+        public void ValidateDate_returns_appropriate_error_if_some_values_invalid(
+            int day,
+            int month,
+            int year,
+            bool dayError,
+            bool monthError,
+            bool yearError,
+            string errorMessageEnding
+        )
+        {
+            // When
+            var result = DateValidator.ValidateDate(day, month, year);
+
+            // Then
+            result.HasDayError.Should().Be(dayError);
+            result.HasMonthError.Should().Be(monthError);
+            result.HasYearError.Should().Be(yearError);
+            result.ErrorMessage.Should().Be("Date must include a real " + errorMessageEnding);
+        }
+
+        [Test]
+        public void ValidateDate_returns_appropriate_error_if_date_not_real()
+        {
+            // When
+            var result = DateValidator.ValidateDate(30, 2, 3000);
+
+            // Then
+            result.HasDayError.Should().BeTrue();
+            result.HasMonthError.Should().BeTrue();
+            result.HasYearError.Should().BeTrue();
+            result.ErrorMessage.Should().Be("Date must be a real date");
+        }
+
+        [Test]
+        public void ValidateDate_returns_appropriate_error_if_date_not_in_future()
+        {
+            // When
+            var result = DateValidator.ValidateDate(1, 1, 2000);
+
+            // Then
+            result.HasDayError.Should().BeTrue();
+            result.HasMonthError.Should().BeTrue();
+            result.HasYearError.Should().BeTrue();
+            result.ErrorMessage.Should().Be("Date must be in the future");
+        }
+
+        [Test]
+        public void ValidateDate_uses_name_correctly_in_error_message()
+        {
+            // When
+            var result = DateValidator.ValidateDate(null, null, null, "What's required", true);
+
+            // Then
+            result.HasDayError.Should().BeTrue();
+            result.HasMonthError.Should().BeTrue();
+            result.HasYearError.Should().BeTrue();
+            result.ErrorMessage.Should().Be("What's required is required");
+        }
+
+        [TestCase(true, false, false)]
+        [TestCase(false, true, false)]
+        [TestCase(false, false, true)]
+        [TestCase(true, true, false)]
+        [TestCase(true, false, true)]
+        [TestCase(false, true, true)]
+        [TestCase(true, true, true)]
+        public void ToValidationResultList_includes_one_error_for_each_erroneous_part_of_date(
+            bool hasDayError,
+            bool hasMonthError,
+            bool hasYearError
+        )

```

Comment:
I think this test has too much logic in it - it's nice to have parameterised tests, but if we need to construct an array with if statements, we're adding logic that technically should be tested itself. Could we simplify this (feel free to break it up into separate tests) such that the additional logic on constructing the errorMemberNames is not required? 🙏